File: Projects/Multithreading/app/src/main/java/com/ivettevaldez/multithreading/common/BaseFragment.java
package com.ivettevaldez.multithreading.common;

import androidx.fragment.app.Fragment;

import com.ivettevaldez.multithreading.common.dependencyinjection.ApplicationCompositionRoot;

public class BaseFragment extends Fragment {

    private ApplicationCompositionRoot applicationCompositionRoot;

    protected final ApplicationCompositionRoot getCompositionRoot() {
        if (applicationCompositionRoot == null) {
            applicationCompositionRoot = ((BaseApplication) requireActivity().getApplication())
                    .getApplicationCompositionRoot();
        }
        return applicationCompositionRoot;
    }
}

File: Projects/UnitTesting/app/src/main/java/com/ivettevaldez/unittesting/tutorialandroidapp/screens/common/controllers/BaseFragment.kt
package com.ivettevaldez.unittesting.tutorialandroidapp.screens.common.controllers

import androidx.fragment.app.Fragment
import com.ivettevaldez.unittesting.tutorialandroidapp.common.CustomApplication
import com.ivettevaldez.unittesting.tutorialandroidapp.common.dependencyinjection.ControllerCompositionRoot

open class BaseFragment : Fragment() {

    private var controllerCompositionRoot: ControllerCompositionRoot? = null

    fun getCompositionRoot(): ControllerCompositionRoot {
        if (controllerCompositionRoot == null) {
            controllerCompositionRoot = ControllerCompositionRoot(
                (requireActivity().application as CustomApplication).compositionRoot,
                requireActivity()
            )
        }
        return controllerCompositionRoot!!
    }
}